File: ServerManagement/src/main/java/org/elkoserver/foundation/servermanagement/DirectorShutdown.java
package org.elkoserver.foundation.servermanagement;

import java.io.IOException;
import java.io.OutputStream;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

import static java.lang.Integer.parseInt;

public class DirectorShutdown {

    public static void main(String[] arguments) throws IOException, InterruptedException {
        if (arguments.length != 2) {
            printUsage();
        } else {
            shutDownServer(arguments[0], parseInt(arguments[1]));
        }
    }

    private static void printUsage() {
        System.out.println("Usage: DirectorShutdown <host address> <port number>");
    }

    private static void shutDownServer(String hostAddress, int portNumber) throws IOException, InterruptedException {
        try (Socket socket = new Socket(hostAddress, portNumber)) {
            OutputStream outputStream = socket.getOutputStream();
            outputStream.write(("{to:\"admin\", op:\"auth\", auth:{mode:\"open\"}}\n\n").getBytes(StandardCharsets.UTF_8));
            Thread.sleep(1_000L);
            outputStream.write(("{to:\"admin\", op:\"shutdown\", director:true, kill:false}\n\n").getBytes(StandardCharsets.UTF_8));
        }
    }
}

File: ServerManagement/src/main/kotlin/org/elkoserver/foundation/servermanagement/DirectorShutdown.kt
package org.elkoserver.foundation.servermanagement

import java.net.Socket

object DirectorShutdown {
    @JvmStatic
    fun main(arguments: Array<String>) {
        startHostPortPasswordProgram("DirectorShutdown", arguments, DirectorShutdown::shutDownServer, true)
    }

    private fun shutDownServer(hostAddress: String, portNumber: Int, password: String) {
        Socket(hostAddress, portNumber).use { socket ->
            val outputStream = socket.getOutputStream()
            outputStream.write("""{"to":"admin", "op":"auth", "auth":{"mode":""".toUtf8())
            if (isNoPasswordIndicator(password)) {
                outputStream.write(""""open"""".toUtf8())
            } else {
                outputStream.write(""""password", "code":"$password"""".toUtf8())
            }
            outputStream.write("""}}$END_OF_COMMAND""".toUtf8())
            Thread.sleep(1000L)
            outputStream.write("""{"to":"admin", "op":"shutdown", "director":true}$END_OF_COMMAND""".toUtf8())
            Thread.sleep(1000L)
        }
    }
}
